### PR TITLE
Rebuild `Context` and Add lifecycle manager

### PR DESCRIPTION
**Pull request by DavdGao**
- The attributes of `ctx` are classified into the following two classes
     - `ctx.xxx`: general attributes
     - `ctx.mode.xxx`: actually refer to `ctx["mode"][ctx.cur_mode]["xxx"]`

 - Add lifecycle manager
     - Build two classes `CtxStatsVar` and `CtxReferVar` to wrap the variables with specific lifecycle and clear function. 
          - `CtxStatsVar`: statistic variables (float), e.g. `loss_batch_total`
               - `lifecycle`: chosen from  `batch`, `epoch`, `routine` and `None` 
          - `CtxReferVar`: reference variable, e.g. `y_prob`, `data_batch`
               - `lifecycle`: chosen from  `batch`, `epoch`, `routine` and `None` 
               - `efunc`: the customized function to run when the lifecycle ends
                   - e.g. for `ctx.model`, maybe we want to move it to cpu when the lifecycle ends. 
                   - When `efunc` is `None`, the variable is deleted at the end of the lifecycle
     - Add an attribute `lifecycles` in `Context` to record the lifecycles of the attributes
     - Add a decorator `@lifecycle(lifecycle=xxx)` to manage  the variables.

### CODE REVIEW

File: federatedscope/cv/dataset/leaf_cv.py [hunk 1]
```diff
@@ -91,6 +91,7 @@ def raw_file_names(self):
     def download(self):
         # Download to `self.raw_dir`.
         url = '/xxx/xxx/share'
+        url = "/mnt/gaodawei.gdw/data/"
```
Comment (yxdyc) [inline federatedscope/cv/dataset/leaf_cv.py:94]: We'd better change the `url` to `cfg.data_path`. @rayrayraykk

File: federatedscope/core/trainers/context.py [hunk 2]
```diff
@@ -1,79 +1,56 @@
...
-
-        loss_total_test (float): accumulated batch loss during test
-        num_samples_test (float): accumulated regular loss during test
-
-        loss_total_val (float): accumulated batch loss during val
-        num_samples_val (float): accumulated regular loss during val
-
-        eval_metrics (dict): evaluation results
-    """
+class BasicDict(dict):
```
Comment (yxdyc) [inline federatedscope/core/trainers/context.py:11]: Please add basic documentation for the new classes

File: federatedscope/core/trainers/context.py [hunk 3]
```diff
@@ -1,79 +1,56 @@
...
 
     __setattr__ = dict.__setitem__
     __delattr__ = dict.__delitem__
 
+    def __init__(self, init_dict=None):
+        if init_dict is not None:
+            super(BasicDict, self).__init__(init_dict)
+        self.lifecycles = collections.defaultdict(set)
+
+    def __setattr__(self, key, value):
```
Comment (yxdyc) [inline federatedscope/core/trainers/context.py:21]: We have re-write the ``__setattr__`` and ``__getattr__`` in lines 16 to 24. Why do we still remain the line 13 to 14?

File: federatedscope/core/trainers/context.py (hunk 2 above)
Comment (yxdyc) [inline federatedscope/core/trainers/context.py:11]: Perhaps the name could be ``LifecycleDict`` ?

File: federatedscope/core/trainers/context.py [hunk 4]
```diff
@@ -185,3 +182,97 @@ def reset_used_dataset(self):
...
+            maintain: if maintain the variable when calling `clear` function
+            lifecycle: specific lifecycle of the attribute
+            efunc: specific the calling function when the lifecycle of the attribute ends
+        """
+        assert lifecycle in BasicCtxVar.LIEFTCYCLES
+
+        self.maintain = maintain
+        self.lifecycle = lifecycle
+
+class CtxReferVar(BasicCtxVar):
```
Comment (yxdyc) [inline federatedscope/core/trainers/context.py:227]: Maybe the name could be ``CtxTempVar``?

**Review by yxdyc — COMMENTED** [review]
Good job! It seems that we can remove some redundant hooks in the child trainers and make the code more elegant! Plz see some minor issues in the in-line comments.

File: federatedscope/cv/dataset/leaf_cv.py (hunk 1 above)
Comment (DavdGao) [reply federatedscope/cv/dataset/leaf_cv.py:94]: This is for local unit tests, I have modified it in the new version

**Comment by DavdGao** [comment]
@yxdyc  The code is modified accordingly.

**Review by Osier-Yi — DISMISSED** [review]

**Comment by yxdyc** [comment]
> @yxdyc The code is modified accordingly.
The [commit](https://github.com/alibaba/FederatedScope/pull/27/commits/e749a3c2d982243b7f16cde101b3e3d98388ca53) looks good to me.

**Review by yxdyc — APPROVED** [review]

**Review by xieyxclack — COMMENTED** [review]
Good job! It looks good to me. And later please update the tutorial accordingly

**Comment by rayrayraykk** [comment]
Looks good to me.

**Review by rayrayraykk — COMMENTED** [review]
Looks good.

File: federatedscope/core/evaluator.py [hunk 5]
```diff
@@ -41,22 +41,20 @@ def eval(self, ctx):
         results = {}
         y_true, y_pred, y_prob = self._check_and_parse(ctx)
         for metric, func in self.eval_metric.items():
-            results["{}_{}".format(ctx.cur_data_split,
-                                   metric)] = func(ctx=ctx,
-                                                   y_true=y_true,
-                                                   y_pred=y_pred,
-                                                   y_prob=y_prob,
-                                                   metric=metric)
+            results["{}_{}".format(ctx.cur_mode, metric)] = func(ctx=ctx,
```
Comment (joneswong) [inline federatedscope/core/evaluator.py:44]: To be discussed: I am curious about how to get the results of "mode \times split"

File: federatedscope/core/trainers/context.py [hunk 6]
```diff
@@ -159,14 +182,21 @@ def pre_calculate_batch_epoch_num(self, local_update_steps):
         return num_train_batch, num_train_batch_last_epoch, num_train_epoch, num_total_train_batch
 
     def append_mode(self, mode):
-        self.mode.append(mode)
-        self.cur_mode = self.mode[-1]
+        if mode in self.mode_stack:
```
Comment (joneswong) [inline federatedscope/core/trainers/context.py:185]: What is the scope of valid modes? According to this code snippet, there are n! modes (suppose there are n choices to be appended)

Comment (DavdGao) [reply federatedscope/core/trainers/context.py:185]: Generally, we choose mode from "train", "test" and "val", which corresponds to the different running processes in FL/ML. 
Actually, only the final mode in `self.mode` represents the current mode, and we track the previous ones to support  nested running routines (e.g. running evaluation within the training process) and recover the status of the learning model (`model.train/eval()`) after jumping out of the nested routine.

**Pull request closed without merging**